### PR TITLE
Remove some uses of singletons in codebase, where feasible

### PR DESCRIPTION
**Pull request by aelred**

### CODE REVIEW

**Comment by aelred** [comment]
!rtg some npm issue?

File: grakn-dist/src/bin/grakn-engine.sh [hunk 1]
```diff
@@ -38,6 +38,9 @@ for jar in "${GRAKN_HOME}"/lib/*.jar; do
     fi
 done
 
+# Add path containing logback.xml
+CLASSPATH="$CLASSPATH":"${GRAKN_HOME}"/conf/main/
```
Comment (alexandraorth) [inline grakn-dist/src/bin/grakn-engine.sh]: This probably should not be hard coded. What if someone specifies their properties and logback conf somewhere else

File: grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java [hunk 2]
```diff
@@ -57,46 +63,48 @@
...
+    private final EngineGraknGraphFactory factory;
+
+    private GraknEngineServer(GraknEngineConfig prop) {
+        this.prop = prop;
+
+        factory = EngineGraknGraphFactory.create(prop.getProperties());
+        taskManager = startTaskManager();
+        usersHandler = UsersHandler.create(prop.getProperty(ADMIN_PASSWORD_PROPERTY), factory);
+
+        Optional<String> secret = prop.tryProperty(JWT_SECRET_PROPERTY);
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java]: We probably should not be doing these static imports.

File: grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java [hunk 3]
```diff
@@ -57,46 +63,48 @@
...
+    private final UsersHandler usersHandler;
+    private final @Nullable JWTHandler jwtHandler;  // TODO: Make sure controllers handle the null case
+    private final EngineGraknGraphFactory factory;
+
+    private GraknEngineServer(GraknEngineConfig prop) {
+        this.prop = prop;
+
+        factory = EngineGraknGraphFactory.create(prop.getProperties());
+        taskManager = startTaskManager();
+        usersHandler = UsersHandler.create(prop.getProperty(ADMIN_PASSWORD_PROPERTY), factory);
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java]: I would move this down to the startHTTP() method because its only used there.

Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java]: Ditto for JWTHandler

File: grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java [hunk 4]
```diff
@@ -120,40 +130,47 @@ private TaskManager startTaskManager(String taskManagerClassName) {
...
-        EngineGraknGraphFactory factory = EngineGraknGraphFactory.getInstance();
         new GraqlController(factory, spark);
         new ConceptController(factory, spark);
         new DashboardController(factory, spark);
-        new SystemController(spark);
-        new AuthController(spark);
-        new UserController(spark);
-        new CommitLogController(spark, taskManager);
+        new SystemController(spark, prop.getProperties());  // TODO: Only pass the necessary properties
+        new AuthController(spark, passwordProtected, jwtHandler, usersHandler);
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java]: Should we only start the `AuthController` if `isPasswordProtected` is true? This would mean that we can instantiate the `JWTHandler` only when it is needed for authentication.

File: grakn-engine/src/main/java/ai/grakn/engine/controller/AuthController.java [hunk 5]
```diff
@@ -92,6 +96,6 @@ private String newSession(Request req, Response res) {
     @ApiOperation(
             value = "Returns true if Engine endpoints are password protected. False otherwise.")
     private String isPasswordProtected(Request req, Response res) {
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/controller/AuthController.java]: This should return a boolean.

File: grakn-engine/src/main/java/ai/grakn/engine/controller/SystemController.java [hunk 6]
```diff
@@ -69,8 +69,12 @@
  */
 public class SystemController {
     private final Logger LOG = LoggerFactory.getLogger(SystemController.class);
+    private final Properties graphProperties;
+    private final EngineGraknGraphFactory factory;
 
-    public SystemController(Service spark) {
+    public SystemController(Service spark, Properties graphProperties) {
+        this.graphProperties = graphProperties;
+        this.factory = EngineGraknGraphFactory.create(graphProperties);
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/controller/SystemController.java]: Can we create a method on the `EngineGraknGraphFactory` that will return the properties? That way we will no have to instantiate another one here.

File: grakn-engine/src/main/java/ai/grakn/engine/session/RemoteSession.java [hunk 7]
```diff
@@ -46,44 +43,46 @@
...
-    @SuppressWarnings("unused")
-    public RemoteSession() {
+    private RemoteSession(@Nullable UsersHandler usersHandler) {
+        this.usersHandler = usersHandler;
     }
 
-    @OnWebSocketConnect
-    public void onConnect(Session session) {
+    public static RemoteSession create() {
+        return new RemoteSession(null);
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/session/RemoteSession.java]: :(

File: grakn-engine/src/main/java/ai/grakn/engine/user/SystemKeyspaceUsers.java [hunk 8]
```diff
@@ -49,6 +49,12 @@
  */
 public class SystemKeyspaceUsers extends UsersHandler {
     private final Logger LOG = LoggerFactory.getLogger(SystemKeyspaceUsers.class);
+    private final EngineGraknGraphFactory factory;
+
+    SystemKeyspaceUsers(String adminPassword, EngineGraknGraphFactory factory) {
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/user/SystemKeyspaceUsers.java]: I don't like that sometimes we pass in this `EngineGraknGraphFactory` and other times we don't.

**Comment by aelred** [comment]
!rtg - looks like there was a timeout? I'm assuming it's a random failure.

File: grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java (hunk 4 above)
Comment (aelred) [inline grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java]: Yes, probably. I'm going to make a TP bug for this and leave it for a different PR - since this bug already existed.

File: grakn-dist/src/bin/grakn-engine.sh (hunk 1 above)
Comment (aelred) [inline grakn-dist/src/bin/grakn-engine.sh]: I agree, but for the future! Ideally all the hard-coded paths in the shell scripts should be defined just once somewhere.

**Comment by aelred** [comment]
!rtg `TitanPreviousPropertyStepTest.whenFilteringAPropertyToBeEqualToAPreviousProperty_UseTitanGraphStep`

File: grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java [hunk 9]
```diff
@@ -120,40 +118,53 @@ private TaskManager startTaskManager(String taskManagerClassName) {
...
         // Start all the controllers
-        EngineGraknGraphFactory factory = EngineGraknGraphFactory.getInstance();
         new GraqlController(factory, spark);
         new ConceptController(factory, spark);
         new DashboardController(factory, spark);
-        new SystemController(spark);
-        new AuthController(spark);
-        new UserController(spark);
-        new CommitLogController(spark, taskManager);
+        new SystemController(factory, spark);
```
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/GraknEngineServer.java]: This actually makes a lot more sense now. It reduces the cohesion of the controller which will help with making the service orientated architecture we have been talking about. In other words. .  . nice.

File: grakn-engine/src/main/java/ai/grakn/engine/controller/CommitLogController.java [hunk 10]
```diff
@@ -49,9 +47,11 @@
  */
 //TODO Implement delete
 public class CommitLogController {
+    private final String defaultKeyspace;
```
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/controller/CommitLogController.java]: I suspect that this can actually be removed. Future PR though as not relevant to the objectives of this PR.

File: grakn-engine/src/main/java/ai/grakn/engine/tasks/manager/singlequeue/SingleQueueTaskManager.java [hunk 11]
```diff
@@ -92,23 +93,28 @@
...
-        this.zookeeper = new ZookeeperConnection();
+    public SingleQueueTaskManager(EngineID engineId, GraknEngineConfig config) {
+        this.config = config;
+        this.zookeeper = new ZookeeperConnection(config);
         this.storage = new TaskStateZookeeperStore(zookeeper);
         this.offsetStorage = new ExternalOffsetStorage(zookeeper);
 
         //TODO check that the number of partitions is at least the capacity
-        this.producer = kafkaProducer();
+        //TODO only pass necessary Kafka properties
```
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/tasks/manager/singlequeue/SingleQueueTaskManager.java]: This TODO should also be whacked on for the engine graph factory properties
